File: 1001-1100/1004-Max Consecutive Ones III/1004-Max Consecutive Ones III.cpp
class Solution {
public:
    int longestOnes(vector<int>& A, int K) {
        int zero = 0, start = 0;
        for (int i = 0; i < A.size(); ++i) {
            if (A[i] == 0) {
                ++zero;
            }

            if (zero > K && A[start++] == 0) {
                --zero;
            }
        }

        return A.size() - start;
    }
};

File: 1001-1100/1004-Max Consecutive Ones III/1004-Max Consecutive Ones III.py
class Solution:
    def longestOnes(self, A: List[int], K: int) -> int:
        maxLen = zero = start = 0
        for i, a in enumerate(A):
            if a == 0:
                zero += 1
            while zero > K:
                if A[start] == 0:
                    zero -= 1
                start += 1
            
            maxLen = max(maxLen, i - start + 1)
        return maxLen


class Solution2:
    def longestOnes(self, A: List[int], K: int) -> int:
        zero = start = 0
        for i, a in enumerate(A):
            if a == 0:
                zero += 1
            if zero > K:
                if A[start] == 0:
                    zero -= 1
                start += 1
            
        return len(A) - start
